File: Reliability/ludecompose.cpp
#include "ludecompose.h"

LUDecompose::LUDecompose(std::vector<std::vector<double> > matrix, std::vector<double> vector, int size)
{
    this->size = size;
    this->vector = std::vector<double>(vector);
    this->matrix = std::vector<std::vector<double> >(matrix);
}

std::vector<double> LUDecompose::lupSolve()
{
    std::vector<std::vector<double> > a1(this->matrix);
    std::vector<double> x(this->size), y(this->size);

    std::vector<int> p(this->size);


    p = lupDecompose(a1);
    y = forwardSub(a1, this->vector, p);
    for (int i = 0; i < y.size(); i++)
    {
        std::cout << this->vector[i] << std::endl;
    }
    x = backSub(a1, y);

    return x;
}

std::vector<int> LUDecompose::lupDecompose(std::vector<std::vector<double>> &mat)
{
    int j, k;
    std::vector<int> p(this->size);
    for (int i = 0; i < this->size; i++)
    {
        p[i] = i;
    }
    for (int i = 0; i < this->size; i++)
    {
        j = MaxInColumn(mat, i);
        if (i != j)
        {
            swapRow(mat, i, j);
            k = p[i];
            p[i] = p[j];
            p[j] = k;
        }
        for (int j = i+1; j < this->size; j++)
        {
            mat[j][i] /= mat[i][i];
        }
        for (int j = i+1; j < this->size; j++)
        {
            for (int k = i+1; k < this->size; k++)
            {
                mat[j][k] -= (mat[i][k]*mat[j][i]);
            }
        }
    }

    return p;
}

void LUDecompose::swapRow(std::vector<std::vector<double> > &mat, int size1, int size2)
{
    auto temp = mat[size1];
    mat[size1] = mat[size2];
    mat[size2] = temp;
}

int LUDecompose::MaxInColumn(std::vector<std::vector<double> > mat, int i) const
{
    int j = i + 1, k = i;

    double max = fabs(mat[i][i]);
    for(; j < this->size; j++)
    {
        if (fabs(mat[j][i]) > max)
        {
            max = fabs(mat[j][i]);
            k = j;
        }
    }

    return k;
}

std::vector<double> LUDecompose::forwardSub(std::vector<std::vector<double> > mat, std::vector<double> vec, std::vector<int> p)
{
    std::vector<double> y(this->size);
    for (int i = 0; i < this->size; i++)
    {
        y[i] = vec[p[i]];
        for (int j = 0; j < i; j++)
        {
            y[i] -= mat[i][j]*y[j];
        }
    }

    return y;
}

std::vector<double> LUDecompose::backSub(std::vector<std::vector<double> > mat, std::vector<double> y)
{
    std::vector<double> x(this->size);

    for (int i = 0; i < this->size; i++)
    {
        x[this->size-i-1] = y[this->size-1-i];
        for (int j = 0; j < i; j++)
        {
            x[this->size-1-i] -= (mat[this->size-1-i][this->size-1-j]*x[this->size-1-j]);
        }
        x[this->size-1-i] /= mat[this->size-1-i][this->size-1-i];
    }

    return x;
}

File: c#/MNK2dim/MNK2dim/LUDecompose.cs
using System;
using System.Collections.Generic; 

namespace MNK
{   
    class LUDecompose
    {
        private int size;
        private double[] vector;
        private double[,] matrix;

        public LUDecompose(double[,] matrix, double[] vector, int size)
        {
            this.size = size;
            this.vector = vector;
            this.matrix = matrix;
        }

        public double[] luSolve()
        {
            double[,] a1;
            double[] y;
            double[] x;

            int[] p;

            a1 = copyMatrix(matrix);




            //for (int i = 0; i < 2; i++)
            //{
            //    for (int j = 0; j < 2; j++)
            //    {
            //        Console.Write(a1[i, j] + " ");
            //    }
            //    Console.WriteLine();
            //}

            p = lupDecompose(ref a1);

            //for (int i = 0; i < 2; i++)
            //{
            //    for (int j = 0; j < 2; j++)
            //    {
            //        Console.Write(a1[i, j] + " ");
            //    }
            //    Console.WriteLine();
            //}
            y = forwardSub(a1, vector, p);
            for (int i = 0; i < y.Length; i++ )
            {
                Console.WriteLine(this.vector[i]);
            }
            x = backSub(a1, y);

            return x;
        }

        private int[] lupDecompose(ref double[,] mat)
        {
            int j, k;
            int[] p = new int[size];

            for (int i = 0; i < size; i++)
            {
                p[i] = i;
            }

            for (int i = 0; i < size; i++)
            {
                j = maxInColumn(mat, i);
                if (i != j)
                {
                    swapRow(mat, i, j);
                    k = p[i];
                    p[i] = p[j];
                    p[j] = k;
                }
                for (j = i+1; j < size; j++)
                {
                    mat[j, i] /= mat[i, i];
                }
                for (j = i+1; j < size; j++)
                {
                    for (k = i + 1; k < size; k++)
                    {
                        mat[j, k] -= (mat[i, k] * mat[j, i]);
                    }
                }
            }

            return p;
        }

        private int maxInColumn(double[,] mat, int i)
        {
            int j = i + 1, k = i;
            double max = Math.Abs(mat[i, i]);
            for (; j < size; j++)
            {
                if (Math.Abs(mat[j, i]) > max)
                {
                    max = Math.Abs(mat[j, i]);
                    k = j;
                }
            }
            return k;
        }

        public void swapRow(double[,] mat, int size1, int size2)
        {
            for (int i = 0; i < size; i++)
            {
                var row = mat[size1, i];
                mat[size1, i] = mat[size2, i];
                mat[size2, i] = row;
            }
        }

        private double[] backSub(double[,] mat, double[] y)
        {
            double[] x = new double[size];

            for (int i = 0; i < size; i++)
            {
                x[size - 1 - i] = y[size - 1 - i];
                for (int j = 0; j < i; j++)
                {
                    x[size - 1 - i] -= (mat[size - 1 - i, size - 1 - j] * x[size - 1 - j]);
                }
                x[size - 1 - i] /= mat[size - 1 - i, size - 1 - i];
            }
            return x;
        }

        private double[] forwardSub(double[,] mat, double[] vec, int[] p)
        { 
            double[] y = new double[size];
            for (int i = 0; i < size; i++)
            {
                y[i] = vec[p[i]];
                for (int j = 0; j < i; j++)
                {
                    y[i] -= mat[i, j] * y[j];
                }
            }
            return y;
        }

        private double[,] copyMatrix(double[,] from)
        {
            double[,] to = new double[from.GetLength(0), from.GetLength(1)];
            for (int i = 0; i < size; i++)
            {
                for (int j = 0; j < size; j++)
                {
                    to[i, j] = from[i, j];
                }
            }
            return to;
        }
    }
}
